lib/DxcSupport/dxcmem.cpp

```diff
@@ -39,97 +37,56 @@ _Ret_maybenull_ _Post_writable_byte_size_(nBytes) void *DxcThreadAlloc(size_t nB
 void DxcThreadFree(void *) throw();
 
 HRESULT DxcInitThreadMalloc() throw() {
+  DXASSERT(g_ThreadMallocTls == nullptr, "else InitThreadMalloc already called");
   DXASSERT(g_pDefaultMalloc == nullptr, "else InitThreadMalloc already called");
 
-  #ifdef _WIN32
-  DXASSERT(g_ThreadMallocTlsIndex == 0, "else InitThreadMalloc already called");
   // We capture the default malloc early to avoid potential failures later on.
   HRESULT hrMalloc = CoGetMalloc(1, &g_pDefaultMalloc);
   if (FAILED(hrMalloc)) return hrMalloc;
 
-  g_ThreadMallocTlsIndex = TlsAlloc();
-  if (g_ThreadMallocTlsIndex == TLS_OUT_OF_INDEXES) {
-    g_ThreadMallocTlsIndex = 0;
-    g_pDefaultMalloc->Release();
-    g_pDefaultMalloc = nullptr;
-    return E_OUTOFMEMORY;
-  }
-  #else
-  g_pDefaultMalloc = new IMalloc;
-  #endif
+  g_ThreadMallocTls = (llvm::sys::ThreadLocal<IMalloc>*)g_pDefaultMalloc->Alloc(sizeof(llvm::sys::ThreadLocal<IMalloc>));
+  g_ThreadMallocTls = new(g_ThreadMallocTls) llvm::sys::ThreadLocal<IMalloc>;
 
   return S_OK;
 }
 
 void DxcCleanupThreadMalloc() throw() {
-  #ifdef _WIN32
-  if (g_ThreadMallocTlsIndex) {
-    TlsFree(g_ThreadMallocTlsIndex);
-    g_ThreadMallocTlsIndex = 0;
-    DXASSERT(g_pDefaultMalloc, "else DxcInitThreadMalloc didn't work/fail atomically");
+  if (g_ThreadMallocTls) {
+    g_ThreadMallocTls->llvm::sys::ThreadLocal<IMalloc>::~ThreadLocal();
```

Comment:
I wonder if you can do
```cpp
delete g_ThreadMallocTls;
```
instead? that should also invoke the destructor, shouldn't it?

Reply:
As you pointed out placement new doesnt really allocate anything. Because of this, it shouldn't be freed using free. Calling the destructor explicitly and then freeing in a way corresponding to the original allocation is the established way to do this and what the compiler expects. There's another example of this in microcom.h.

Reply:
oh yes. You're right.